pulsar-io/hbase/src/main/java/org/apache/pulsar/io/hbase/sink/HbaseAbstractSink.java

```diff
@@ -0,0 +1,205 @@
+    private Configuration configuration;
+    private Connection connection;
+    private Admin admin;
+    private TableName tableName;
+    private Table table;
+
+    protected TableDefinition tableDefinition;
+
+    // for flush
+    private long batchTimeMs;
+    private int batchSize;
+    private List<Record<T>> incomingList;
+    private ScheduledExecutorService flushExecutor;
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+        hbaseSinkConfig = HbaseSinkConfig.load(config);
+        Preconditions.checkNotNull(hbaseSinkConfig.getZookeeperQuorum(), "zookeeperQuorum property not set.");
+        Preconditions.checkNotNull(hbaseSinkConfig.getZookeeperClientPort(), "zookeeperClientPort property not set.");
+        Preconditions.checkNotNull(hbaseSinkConfig.getZookeeperZnodeParent(), "zookeeperZnodeParent property not set.");
+        Preconditions.checkNotNull(hbaseSinkConfig.getTableName(), "hbase tableName property not set.");
+
+        getTable(hbaseSinkConfig);
+        tableDefinition = getTableDefinition(hbaseSinkConfig);
+
+        batchTimeMs = hbaseSinkConfig.getBatchTimeMs();
+        batchSize = hbaseSinkConfig.getBatchSize();
+        incomingList = Lists.newArrayList();
+        flushExecutor = Executors.newScheduledThreadPool(1);
+        flushExecutor.scheduleAtFixedRate(() -> flush(), batchTimeMs, batchTimeMs, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (null != admin) {
+            admin.close();
+        }
+
+        if (null != connection) {
+            connection.close();
+        }
+
+        if (null != flushExecutor) {
+            flushExecutor.shutdown();
+        }
+    }
+
+    @Override
+    public void write(Record<T> record) throws Exception {
+        int number;
+        synchronized (this) {
+            if (null != record) {
+                incomingList.add(record);
+            }
+            number = incomingList.size();
+        }
+
+        if (number == batchSize) {
+            flushExecutor.submit(() -> flush());
+        }
+    }
+
+    private void flush() {
+        List<Put> puts = new ArrayList<>();
+        List<Record<T>>  toFlushList;
+        synchronized (this) {
+            if (incomingList.isEmpty()) {
+                return;
+            }
+            toFlushList = incomingList;
+            incomingList = Lists.newArrayList();
+        }
+
+        if (CollectionUtils.isNotEmpty(toFlushList)) {
+            for (Record<T> record: toFlushList) {
+                try {
+                    bindValue(record, puts);
+                } catch (Exception e) {
+                    log.warn("Record flush thread was exception ", e);
+                    toFlushList.remove(record);
+                }
+            }
+        }
+
+        try {
+            if (CollectionUtils.isNotEmpty(puts)) {
+                table.put(puts);
```

Comment:
> I am not sure why line 158-159 is not in the `if` block. because if the `puts` is not put to the table, there is no point for you to call `admin.flush`, right?

yes